### ISSUE TITLE
API Proposal: HD Archiving

### ISSUE THREAD

**Issue by msach22** [opentok/Opentok-.NET-SDK#101]
This is an API Proposal regarding the recently added platform functionality for archiving in HD. 

## Archive Resolution
The platform now allows the resolution to be set for Composed archives. See [docs](https://tokbox.com/developer/rest/#start_archive)
The parameter is a string: `resolution` -  The resolution of the archive (either "640x480" or "1280x720"). This property is only set for composed archives.

We propose to add a new optional parameter to the `Opentok.StartArchive()` method for setting the resolution.

### `StartArchive()`
```C#
# Default mode for starting a composed archive. If a resolution is not provided, the resolution for composed archives will default to "640x480". 
var archive = OpenTok.StartArchive(sessionId);

# To see which resolution was set, the developer can do the following:
var archiveResolution = archive.Resolution;

# Usage for HD archiving
var archive = StartArchive(sessionId, resolution: "1280x720")

# Default mode for starting an individual archive.
var archive = StartArchive(sessionId, outputMode: OutputMode.INDIVIDUAL)

# Setting the resolution for individual mode archiving is not allowed and will throw OpenTokArgumentException
var archive = StartArchive(sessionId, outputMode: OutputMode.INDIVIDUAL, resolution: "1280x720")

# For the individual archiving mode, the Resolution getter would result in a null value.
```

**Comment by utkarshkhokhar** [comment]
LGTM 👍

**Comment by robjperez** [comment]
I don't like that a given combination of parameters could produce an error (probably the user won't even capture the error). IMHO, it is better if we can limit that, so the user doesn't need to worry if the parameters are ok or not.

My proposal is to modifify the outputMode to include the resolution, so there won't be a way to specify INDIVIDUAL + HD.

Something like this:
```js
var archive = StartArchive(sessionId, outputMode: OutputMode.INDIVIDUAL_VGA)

var archive = StartArchive(sessionId, outputMode: OutputMode.COMPOSED_VGA)
var archive = StartArchive(sessionId, outputMode: OutputMode.COMPOSED_HD)

# This will produce a compilation error since there isn't INDIVIDUAL_HD in OutputMode
var archive = StartArchive(sessionId, outputMode: OutputMode.INDIVIDUAL_HD) 
```

**Comment by msach22** [comment]
@robjperez I personally think that this adds complexity and is not as clear as specifying a resolution. This would also change the current options for `OutputMode` instead of introducing another option (resolution) which is what's actually changing. The term `HD` can be confusing if in the future the platform supports (1280x960) which is high definition with a 4:3 aspect ratio in the future.

**Comment by msach22** [comment]
@robjperez This is a good point, I'm okay with adding to the existing enum for `outputMode`, but I don't think we should assume that the caller of the API doesn't care about the exact resolution.

We could do something like `OutputMode.COMPOSED_1280x720` so we don't mislabel the resolution.

@jpujol What do you think? @utkarshkhokhar ?

**Comment by msach22** [comment]
@robjperez I've been thinking about this more and I prefer a resolution parameter over adding the resolution as a part of the `outputMode`. Here's why:

- The JS, PHP, & Python SDKs follow the same approach and it's good to have the same parity because it's easier for developers to go from one SDK to another.
- The REST API also has `outputMode` and `resolution` properties instead of having both be a part of the `outputMode`.

I do think that we can have enums instead of strings for the resolution parameter following Apple's example that you linked. We would also raise an exception if individual mode and resolution were provided instead of letting the request go through and having the API server respond with a bad request response.

**Comment by robjperez** [comment]
It is try that we need to be consistent in all languages, but I think the way to go is to change the others instead of following a bad pattern just for consistency.

About the API, I think a REST api is a different thing than a SDK. You cannot enforce types like enums (there is no compilation stage) in http request so answering a `Bad Request` is acceptable if the user of the API hasn't read the documentation.

**Comment by msach22** [comment]
I respectfully disagree. I don't think that this is a "bad" pattern because the same argument can be made that the developer would get an OpenTokArgumentException if they don't read the documentation for the Server SDKs.

I would prefer not to break the three SDKs for the hundreds of partners that are using our SDKs.

**Comment by robjperez** [comment]
You don't need to break the SDK to add new values to an enum, old values are still valid so no code will be broken with this change.

When I was saying that this method is like setting a trap for the developer is just because of that, because the developer will need to read the documentation (do you remember Uncle Bob about the code is the documentation?) when we can help them by using the compiler. Moreover, some developers will fill the parameter using a variable so we are forcing them to put an `if` to be sure that the combination of parameters is valid AND surronding everything with a `try/catch`, so our lazyness ;) are adding more work to our beloved developers. (What would you put in that catch?)

I will put you an example. Let's say we have a method `orderApples(double numberOfApples)`.  Then a developer just call it like this `orderApples(1.2)` and then we fail saying `You cannot order fractional number of apples`. If I'm the user of that method I would ask me... why don't they just declare the method with an `int` instead of an `double`? Using your argument, If I'm the developer of the sdk, I can say.. hey it is clearly stated in the documentation that you can just order a whole number of Apples.

This kind of traps remind me to old Win32 times:
```
CreateFile(
       L".\\temp.txt",
       GENERIC_READ, 
       NULL,
       NULL,
       CREATE_NEW,
       FILE_ATTRIBUTE_NORMAL,
       NULL);
)
```
I always wonder why I needed to set half of parameters to `NULL`, and how many headaches I had because i misplaced one of the paremeters, but hey, it is well explained in the documentation ;)

**Comment by puneetshettyp** [comment]
Very compelling arguments on both sides. I do not want to comment on usability of the method. But I do want to comment on how the resolution parameter is expected to evolve in the future. We have started an API review for web-view based layout control for archiving. The idea with this advanced layout control would be to allow for more flexible layouts and hence move beyond the standard resolutions for some of the use-cases. I still do not know how the new API will take shape but it would be good to keep this in mind so that what we implement is extensible in the future.

**Comment by robjperez** [comment]
@puneetshettyp I think we need to control the resolution in a way or the other. We cannot allow arbitrary values to be passed as argument, or we could call the method using `"1x1"` or `"1234x9876"` values that we will probably fail.

I think using enum values is extensible in the sense that you can always add new values to the enum in future versions of the SDK.

**Comment by jpujol** [comment]
I would not use `HD` or names for resolutions... in the future we may have more, and putting names to all of them is going to be confusing

**Comment by jeffswartz** [comment]
I like this proposal. I suppose we could add enum values for the two supported resolutions, but I don't see it as necessary (and I agree with @jpujol's point about naming possibly conflicting with future resolutions). I think we should have the implementation in the OpenTok .NET SDK match the REST API and the implementation in the other server SDKs (and this proposal matches those).

I do not like the proposal for`OutputMode.COMPOSED_VGA`, etc. For one thing, it would be a breaking change (we already have `OutputMode.COMPOSED`, which would have to go away, I suppose. Also, the existing `OpenTok.CreateSession()` method can take conflicting parameters (and result in an error, I suppose): `MediaMode.RELAYED` and `ArchiveMode.ALWAYS`. So, I prefer separate parameters for `outputMode` and `resolution`, as is proposed here.

In the future, if we allow more resolutions (or arbitrary resolutions), we could add support for passing an `ArchiveResolution` object that has `Width` and `Height` properties. (And we could deprecate passing in `resolution` as a string. In the .NET SDK and in the other server-side SDKs.)

**Comment by robjperez** [comment]
@jeffswartz It isn't a breaking change, we can deprecate `OutputMode.COMPOSED` and say that it will be same that using `OutputMode.COMPOSED_VGA` so we can support both elements in the enum.
If you read Apple documentation about `AVCaptureSession.Preset`, it is exactly the same.

In general I think having functions with conflicting parameters is not a good idea, there are always a workaround to them. For the CreateSession, we could have two versions of the CreateSession: `CreateSessionRelayed` and `CreateSessionRouted` and both could have different parameters so you cannot end with a wrong combination of arguments. I still think that a REST endpoint is not a programming language SDK where you have a compilation stage to detect problems. If not, why don't just we accept a Hash as a single parameter with all the options bundled in? :)

**Comment by msach22** [comment]
@robjperez A couple of points here:
- I don't think introducing two methods to create sessions is the right approach because it's more confusing than helpful.
- Secondly, I don't think having a resolution parameter with arbitrary resolution values is a bad thing because it gives the developer the luxury to set whichever resolution they want. Just because the developer has the option to feed in a custom resolution, it does not mean that they will feed in anything that's useless to them because it is a feature they pay for.

I would like to close this API review by this Wednesday so if anyone else has any comments, please add them below because we will begin implementation this Wednesday. cc: @jeffswartz @puneetshettyp @jpujol

**Comment by jpujol** [comment]
@robjperez we may accept any resolution in the future, up to a certain value. I think the enum solution will not scale

**Comment by robjperez** [comment]
@jpujol when you say _any_, you mean an arbitrary string?, Would I able to use "1234x2" or "0.5x1000" ?

**Comment by jpujol** [comment]
well, any *valid* resolution, from a minimum to maybe 2k. So valid resolutions would be 400x300, 100x100 or 1280x960, or even in portrait mode: 300x400, 960x1280.

Probably the minimum value for any dimension will be 16 (just taking a guess here)

**Comment by msach22** [comment]
This has been released with `v3.1.0`.
